src/v/cluster/drain_manager.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2022 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+#include "cluster/fwd.h"
+#include "seastarx.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/util/log.hh>
+
+#include <chrono>
+
+namespace cluster {
+
+/*
+ * The drain manager is responsible for managing the draining of leadership from
+ * a node. It is a core building block for implementing node maintenance mode.
+ */
+class drain_manager : public ss::peering_sharded_service<drain_manager> {
+    static constexpr size_t max_parallel_transfers = 25;
+    static constexpr std::chrono::duration transfer_throttle
+      = std::chrono::seconds(5);
+
+public:
+    /*
+     * finsihed:     draining has completed
+     * errors:       draining finished with errors
+     * partitions:   total partitions
+     * eligible:     total drain-eligible partitions
+     * transferring: total partitions currently transferring
+     * failed:       total transfers failed in last batch
+     *
+     * the optional fields may not be set if draining has been requested, but
+     * not yet started. in this case the values are not yet known.
+     */
+    struct drain_status {
+        bool finished{false};
+        bool errors{false};
+        std::optional<size_t> partitions;
+        std::optional<size_t> eligible;
+        std::optional<size_t> transferring;
+        std::optional<size_t> failed;
+    };
+
+    explicit drain_manager(ss::sharded<cluster::partition_manager>&);
+
+    ss::future<> start();
+    ss::future<> stop();
+
+    /*
+     * Start draining this broker.
+     *
+     * Invoke this on each core.
+     */
+    ss::future<> drain();
+
+    /*
+     * Restore broker to a non-drain[ing] state.
+     *
+     * Invoke this on each core.
+     */
+    ss::future<> restore();
```

Comment:
Any reason not to encapsulate the invocation on every core?

Reply:
hmm, not exactly. just trying to write things so it would be easier to unit test without sticking it in a sharded<T>. but I think I messed that up because the status() does a global map_reduce. probably should have made that a free function :/